packages/@aws-cdk/pipelines/README.md

```diff
@@ -62,6 +107,524 @@ class MyApplication extends Stage {
+### Synth and sources
+
+To define a pipeline, instantiate a `CodePipeline` construct. It takes
+one argument, a `synth` step, which is expected to produce the CDK Cloud
+Assembly as its single output (the contents of the `cdk.out` directory after
+running `cdk synth`). "Steps" are arbitrary actions in the pipeline, typically
+used to run scripts or commands.
+
+For the synth, use a `ScriptStep` and specify the commands necessary to build
+your project and run `cdk synth`; the specific commands required will depend on
+the programming language you are using. For a typical NPM-based project, the synth
+will look like this:
+
+```ts
+const source = /* the repository source */;
+
+const pipeline = new CodePipeline(this, 'Pipeline', {
+  synth: new ScriptStep('Synth', {
+    input: source,
+    commands: [
+      'npm ci',
+      'npm run build',
+      'npx cdk synth',
+    ],
+  }),
+});
+```
+
+The pipeline assumes that your `ScriptStep` will produce a `cdk.out`
+directory in the root, containing the CDK cloud assembly. If your
+CDK project lives in a subdirectory, be sure to adjust the
+`primaryOutputDirectory` to match:
+
+```ts
+const pipeline = new CodePipeline(this, 'Pipeline', {
+  synth: new ScriptStep('Synth', {
+    input: source,
+    commands: [
+      'cd mysubdir',
+      'npm ci',
+      'npm run build',
+      'npx cdk synth',
+    ],
+    primaryOutputDirectory: 'mysubdir/cdk.out',
+  }),
+});
+```
+
+The underlying CodePipeline `Pipeline` construct will be produced
+when `app.synth()` is called. You can also force it to be produced
+earlier by calling `pipeline.buildPipeline()`. After you've called
+that method, you can inspect the constructs that were produced by
+accessing the properties of the `pipeline` object.
+
+#### CodePipeline Sources
+
+In CodePipeline, *Sources* define where the source of your application lives.
+When a change to the source is detected, the pipeline will start executing.
+Sources objects can be created by factory methods on the `CodePipelineSource` class:
+
+* `CodePipelineSource.gitHub('org/repo', { branch, authentication })`: use a
+  GitHub respository as the source. By default, it assumes your GitHub OAuth
+  token lives in a Secret in Secrets Manager with the name `github-token`.
+  You can override that by passing the `authentication` property.
+
+#### Additional inputs
+
+`ScriptStep` allows passing in more than one input: additional
+inputs will be placed in the directories you specify. Any step that produces an
+output file set can be used as an input, such as a `CodePipelineSource`, but
+also other `ScriptSteps`:
+
+```ts
+const prebuild = new ScriptStep('Prebuild', {
+  input: CodePipelineSource.gitHub('myorg/repo1'),
+  primaryOutputDirectory: './build',
+  commands: ['./build.sh'],
+});
+
+const synth = new ScriptStep('Synth', {
+  input: CodePipelineSource.gitHub('myorg/repo2'),
+  additionalInputs: {
+    'subdir': CodePipelineSource.gitHub('myorg/repo3'),
+    '../siblingdir': prebuild,
```

Comment:
What? Does that mean it will place an these artifacts outside of the working directory? Sounds like we should actually prevent that, no? Can I do `../../../../usr/lib` here?